### PR TITLE
(GH742) Added TravisCI build system support

### PR DESCRIPTION
**Pull request by RLittlesII**
I am marking this WIP because I am looking for feedback on [Travis Fold](https://github.com/cake-build/cake/compare/develop...RLittlesII:GH742?expand=1#diff-c5e51703b71fb865ef84a947b9548557R10) and the [Write Message](https://github.com/cake-build/cake/compare/develop...RLittlesII:GH742?expand=1#diff-60795439dcfc3f16aa5dc9ab99ac75ddR119) format. Also, are there any other Travis functions we want to expose besides fold?

I will update with all the environment information that was required in #742 

- [x] Travis Fold

- [x] Travis Environment Information

- [x] Unit Tests

### CODE REVIEW

**Comment by RLittlesII** [comment]
@devlead @patriksvensson I feel this is ready for review.

File: src/Cake.Common.Tests/Unit/Build/TravicCI/Data/TravisJobInfoTests.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
+﻿using Cake.Common.Tests.Fixtures.Build;
+using Xunit;
+
+namespace Cake.Common.Tests.Unit.Build.TravicCI.Data
+{
+    public class TravisJobInfoTests
```
Comment (patriksvensson) [inline src/Cake.Common.Tests/Unit/Build/TravicCI/Data/TravisJobInfoTests.cs]: Should be `sealed` for consistency.

File: src/Cake.Common.Tests/Unit/Build/TravicCI/TravisCIEnvironmentInfoTests.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+                // When
+                var result = info.Home;
+
+                // Then
+                Assert.Equal("/home/travis", result);
+            }
+        }
+
+        public sealed class ThtTravisProperty
```
Comment (patriksvensson) [inline src/Cake.Common.Tests/Unit/Build/TravicCI/TravisCIEnvironmentInfoTests.cs]: Misspelling: `Tht`

File: src/Cake.Common/Build/TravisCI/Data/TravisCIJobInfo.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
+﻿using Cake.Core;
+
+namespace Cake.Common.Build.TravisCI.Data
+{
+    /// <summary>
+    /// Provides Travis CI job information for the current build.
+    /// </summary>
+    public class TravisCIJobInfo : TravisCIInfo
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIJobInfo.cs]: Should be `sealed`.

File: src/Cake.Common/Build/TravisCI/Data/TravisCIEnvironmentInfo.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
+﻿using Cake.Core;
+
+namespace Cake.Common.Build.TravisCI.Data
+{
+    /// <summary>
+    /// Provides TravisCI environment information for the current build.
+    /// </summary>
+    public class TravisCIEnvironmentInfo : TravisCIInfo
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIEnvironmentInfo.cs]: Should be `sealed`.

File: src/Cake.Common/Build/TravisCI/Data/TravisCIBuildInfo.cs [hunk 5]
```diff
@@ -0,0 +1,95 @@
+﻿using Cake.Core;
+
+namespace Cake.Common.Build.TravisCI.Data
+{
+    /// <summary>
+    /// Provides Travis CI build information for the current build.
+    /// </summary>
+    public class TravisCIBuildInfo : TravisCIInfo
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIBuildInfo.cs]: Should be `sealed`.

File: src/Cake.Common/Build/TravisCI/Data/TravisCIEnvironmentInfo.cs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        public bool Travis
+        {
+            get { return GetEnvironmentBoolean("TRAVIS"); }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TravisCIEnvironmentInfo"/> class.
+        /// </summary>
+        /// <param name="environment">The environment.</param>
+        public TravisCIEnvironmentInfo(ICakeEnvironment environment) : base(environment)
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIEnvironmentInfo.cs]: New line before `:`

File: src/Cake.Common/Build/TravisCI/Data/TravisCIJobInfo.cs [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+        public bool SecureEnvironmentVariables
+        {
+            get { return GetEnvironmentBoolean("TRAVIS_SECURE_ENV_VARS"); }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TravisCIJobInfo"/> class.
+        /// </summary>
+        /// <param name="environment">The environment.</param>
+        public TravisCIJobInfo(ICakeEnvironment environment) : base(environment)
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIJobInfo.cs]: New line before `:`

File: src/Cake.Common/Build/TravisCI/Data/TravisCIRepositoryInfo.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
+﻿using Cake.Core;
+
+namespace Cake.Common.Build.TravisCI.Data
+{
+    /// <summary>
+    /// Provides TravisCI repository information for the current build.
+    /// </summary>
+    public class TravisCIRepositoryInfo : TravisCIInfo
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIRepositoryInfo.cs]: Should be `sealed`.

File: src/Cake.Common/Build/TravisCI/Data/TravisCIRepositoryInfo.cs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        public string Slug
+        {
+            get { return GetEnvironmentString("TRAVIS_REPO_SLUG"); }
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TravisCIRepositoryInfo"/> class.
+        /// </summary>
+        /// <param name="environment">The environment.</param>
+        public TravisCIRepositoryInfo(ICakeEnvironment environment) : base(environment)
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIRepositoryInfo.cs]: New line before `:`.

File: src/Cake.Common/Build/TravisCI/ITravisCIProvider.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+namespace Cake.Common.Build.TravisCI
+{
+    /// <summary>
+    /// Represents a TravisCI Provider.
+    /// </summary>
+    public interface ITravisCIProvider
+    {
+        /// <summary>
+        /// Gets a value indicating whether this instance is running on travis ci.
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/ITravisCIProvider.cs]: Should be spelled `TravisCI` right?

File: src/Cake.Common/Build/TravisCI/ITravisCIProvider.cs [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+        /// </summary>
+        /// <value>
+        /// <c>true</c> if this instance is running on travis ci; otherwise, <c>false</c>.
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/ITravisCIProvider.cs]: Should be spelled `TravisCI` right?

File: src/Cake.Common/Build/TravisCI/TravisCIInfo.cs [hunk 12]
```diff
@@ -0,0 +1,66 @@
+﻿using System;
+using Cake.Core;
+
+namespace Cake.Common.Build.TravisCI
+{
+    /// <summary>
+    /// Base class used to provide information about the TravisCI environment.
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/TravisCIInfo.cs]: Should be spelled `Travis CI` right?

File: src/Cake.Common.Tests/Unit/Build/BuildSystemAliasesTests.cs [hunk 13]
```diff
@@ -79,5 +79,18 @@ public void Should_Throw_If_Context_Is_Null()
                 Assert.IsArgumentNullException(result, "context");
             }
         }
+        
+        public sealed class TheTravisCIMethod
+        {
+            [Fact]
+            public void Should_Throw_If_Context_Is_Null()
+            {
+                //Given, When
```
Comment (patriksvensson) [inline src/Cake.Common.Tests/Unit/Build/BuildSystemAliasesTests.cs]: Space after `//`.

File: src/Cake.Common.Tests/Unit/Build/BuildSystemAliasesTests.cs [hunk 14]
```diff
@@ -79,5 +79,18 @@ public void Should_Throw_If_Context_Is_Null()
...
+                var result = Record.Exception(() => BuildSystemAliases.TravisCI(null));
+
+                //Then
```
Comment (patriksvensson) [inline src/Cake.Common.Tests/Unit/Build/BuildSystemAliasesTests.cs]: Space after `//`.

File: src/Cake.Common/Build/BuildSystem.cs [hunk 15]
```diff
@@ -33,7 +35,8 @@ public sealed class BuildSystem
         /// <param name="continuaCIProvider">The Continua CI Provider.</param>
         /// <param name="jenkinsProvider">The Jenkins Provider.</param>
         /// <param name="bitriseProvider">The Bitrise Provider.</param>
-        public BuildSystem(IAppVeyorProvider appVeyorProvider, ITeamCityProvider teamCityProvider, IMyGetProvider myGetProvider, IBambooProvider bambooProvider, IContinuaCIProvider continuaCIProvider, IJenkinsProvider jenkinsProvider, IBitriseProvider bitriseProvider)
+        /// <param name="travisCIProvider">The travis ci provider.</param>
```
Comment (patriksvensson) [inline src/Cake.Common/Build/BuildSystem.cs]: Should be spelled `Travis CI`.

File: src/Cake.Common/Build/BuildSystem.cs [hunk 16]
```diff
@@ -336,6 +344,37 @@ public IBitriseProvider Bitrise
...
+        public bool IsRunningOnTravisCI
+        {
+            get { return _travisCIProvider.IsRunningOnTravisCI; }
+        }
+
+        /// <summary>
+        /// Gets the TravisCI provider.
+        /// </summary>
+        /// <value>
+        /// The travis ci.
```
Comment (patriksvensson) [inline src/Cake.Common/Build/BuildSystem.cs]: Should be spelled `Travis CI`.

File: src/Cake.Common/Build/BuildSystem.cs [hunk 17]
```diff
@@ -336,6 +344,37 @@ public IBitriseProvider Bitrise
...
+        /// <value>
+        /// <c>true</c> if this instance is running on travis ci; otherwise, <c>false</c>.
+        /// </value>
+        public bool IsRunningOnTravisCI
+        {
+            get { return _travisCIProvider.IsRunningOnTravisCI; }
+        }
+
+        /// <summary>
+        /// Gets the TravisCI provider.
```
Comment (patriksvensson) [inline src/Cake.Common/Build/BuildSystem.cs]: Should be spelled `Travis CI`.

File: src/Cake.Common/Build/TravisCI/Data/TravisCIBuildInfo.cs [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+        /// <value>
+        /// The build number.
+        /// </value>
+        public int BuildNumber
+        {
+            get { return GetEnvironmentInteger("TRAVIS_BUILD_NUMBER"); }
+        }
+        
+        /// <summary>
+        /// Gets a value indicating whether [travis secure environment variables].
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIBuildInfo.cs]: Comment seems autogenerated by GhostDoc or similar.

File: src/Cake.Common/Build/TravisCI/Data/TravisCIBuildInfo.cs [hunk 19]
```diff
@@ -0,0 +1,95 @@
...
+        /// </summary>
+        /// <value>
+        /// <c>true</c> if [travis secure environment variables]; otherwise, <c>false</c>.
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIBuildInfo.cs]: Comment seems autogenerated by GhostDoc or similar.

File: src/Cake.Common/Build/TravisCI/Data/TravisCIEnvironmentInfo.cs [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+    /// Provides TravisCI environment information for the current build.
+    /// </summary>
+    public class TravisCIEnvironmentInfo : TravisCIInfo
+    {
+        private readonly TravisCIBuildInfo _buildProvider;
+        private readonly TravisCIJobInfo _jobProvider;
+        private readonly TravisCIRepositoryInfo _repositoryProvider;
+
+        /// <summary>
+        /// Gets TravisCI build information.
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIEnvironmentInfo.cs]: Should be `Travis CI`.

File: src/Cake.Common/Build/TravisCI/Data/TravisCIEnvironmentInfo.cs [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+        /// <value>
+        /// The build.
+        /// </value>
+        public TravisCIBuildInfo Build
+        {
+            get { return _buildProvider; }
+        }
+
+        /// <summary>
+        /// Gets TravisCI job information.
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIEnvironmentInfo.cs]: Should be `Travis CI`.

File: src/Cake.Common/Build/TravisCI/Data/TravisCIEnvironmentInfo.cs [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+        /// <value>
+        /// The job.
+        /// </value>
+        public TravisCIJobInfo Job
+        {
+            get { return _jobProvider; }
+        }
+
+        /// <summary>
+        /// Gets TravisCI repository information.
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIEnvironmentInfo.cs]: Should be `Travis CI`.

File: src/Cake.Common/Build/TravisCI/Data/TravisCIEnvironmentInfo.cs [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+        /// <value>
+        /// The repository.
+        /// </value>
+        public TravisCIRepositoryInfo Repository
+        {
+            get { return _repositoryProvider; }
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether this <see cref="TravisCIEnvironmentInfo" /> is ci.
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIEnvironmentInfo.cs]: Comment seems autogenerated by GhostDoc.

File: src/Cake.Common/Build/TravisCI/Data/TravisCIEnvironmentInfo.cs [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+        /// <value>
+        /// <c>true</c> if ci; otherwise, <c>false</c>.
+        /// </value>
+        public bool CI
+        {
+            get { return GetEnvironmentBoolean("CI"); }
+        }
+
+        /// <summary>
+        /// Gets the TravisCI home directory.
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIEnvironmentInfo.cs]: Should be `Travis CI`.

File: src/Cake.Common/Build/TravisCI/Data/TravisCIEnvironmentInfo.cs [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+        /// <value>
+        /// The home.
+        /// </value>
+        public string Home
+        {
+            get { return GetEnvironmentString("HOME"); }
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether the environment is travis.
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIEnvironmentInfo.cs]: Should be `Travis CI`.

File: src/Cake.Common/Build/TravisCI/Data/TravisCIJobInfo.cs [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+        /// <value>
+        /// The name of the os.
+        /// </value>
+        public string OSName
+        {
+            get { return GetEnvironmentString("TRAVIS_OS_NAME"); }
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether [secure environment variables].
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIJobInfo.cs]: Autogenerated by GhostDoc or similar?

File: src/Cake.Common/Build/TravisCI/Data/TravisCIJobInfo.cs [hunk 27]
```diff
@@ -0,0 +1,62 @@
...
+        /// </summary>
+        /// <value>
+        /// <c>true</c> if [secure environment variables]; otherwise, <c>false</c>.
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIJobInfo.cs]: Autogenerated by GhostDoc or similar?

File: src/Cake.Common/Build/TravisCI/Data/TravisCIRepositoryInfo.cs [hunk 28]
```diff
@@ -0,0 +1,62 @@
+﻿using Cake.Core;
+
+namespace Cake.Common.Build.TravisCI.Data
+{
+    /// <summary>
+    /// Provides TravisCI repository information for the current build.
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/Data/TravisCIRepositoryInfo.cs]: Should be `Travis CI`.

File: src/Cake.Common/Build/TravisCI/ITravisCIProvider.cs [hunk 29]
```diff
@@ -0,0 +1,38 @@
+﻿using Cake.Common.Build.TravisCI.Data;
+
+namespace Cake.Common.Build.TravisCI
+{
+    /// <summary>
+    /// Represents a TravisCI Provider.
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/ITravisCIProvider.cs]: Should be `Travis CI`.

File: src/Cake.Common/Build/TravisCI/ITravisCIProvider.cs (hunk 11 above)
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/ITravisCIProvider.cs]: Should be `Travis CI`.

File: src/Cake.Common/Build/TravisCI/TravisCIDisposableExtensions.cs [hunk 30]
```diff
@@ -0,0 +1,54 @@
...
+        /// </summary>
+        internal sealed class TravisCIActionDisposable : IDisposable
+        {
+            private readonly Action<ITravisCIProvider> _disposeAction;
+            private readonly ITravisCIProvider _travisCiProvider;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="TravisCIActionDisposable"/> class.
+            /// </summary>
+            /// <param name="travisCiProvider">The travis ci provider.</param>
```
Comment (patriksvensson) [inline src/Cake.Common/Build/TravisCI/TravisCIDisposableExtensions.cs]: Should be `Travis CI`.

**Comment by patriksvensson** [comment]
@RLittlesII Looks good to me. Left some comments but they're mostly about consistency in naming and code style and should be easy to fix. Great job! :+1:

**Comment by RLittlesII** [comment]
Cleaned up and rebased.

**Comment by gep13** [comment]
@RLittlesII sorry to be a pain, but can you rebase again? Thanks!

**Comment by devlead** [comment]
Thanks for rebasing @RLittlesII :+1: I'll merge.

**Comment by devlead** [comment]
Merged! @RLittlesII thanks for contributing :+1:

**Pull request merged**